### PR TITLE
[Compression - 8] Enable reader to read from compressed files/messages

### PR DESCRIPTION
**Pull request by piraka9011**
This PR allows a Reader to decompress a file or serialized bag message by checking the bag file metadata.
It currently supports the Zstd decompressor only.
Creating a decompressor factory will be part of a later change.

## Changes

* Reader can now read from a compressed file or message.
* Validate metadata keys for compression, supporting older metadata versions.
* Unit test to check that compression occurs when specified in metadata.
* Unit test to check that compression options can be converted b/w string/enum.

Part of [aws-roadmap/#68](https://github.com/ros-security/aws-roadmap/issues/68).

### CODE REVIEW

File: rosbag2_cpp/include/rosbag2_cpp/readers/sequential_reader.hpp [hunk 1]
```diff
@@ -125,10 +127,27 @@ class ROSBAG2_CPP_PUBLIC SequentialReader
   std::shared_ptr<SerializationFormatConverterFactoryInterface> converter_factory_{};
   std::shared_ptr<rosbag2_storage::storage_interfaces::ReadOnlyInterface> storage_{};
   std::unique_ptr<Converter> converter_{};
+  std::unique_ptr<rosbag2_compression::BaseDecompressorInterface> decompressor_;
```
Comment (zmichaels11) [inline rosbag2_cpp/include/rosbag2_cpp/readers/sequential_reader.hpp]:
```c++
std::unique_ptr<rosbag2_compression::BaseDecompressorInterface> decompressor_{};
```

File: rosbag2_cpp/include/rosbag2_cpp/readers/sequential_reader.hpp [hunk 2]
```diff
@@ -125,10 +127,27 @@ class ROSBAG2_CPP_PUBLIC SequentialReader
...
   std::unique_ptr<rosbag2_storage::MetadataIo> metadata_io_{};
   rosbag2_storage::BagMetadata metadata_{};
   std::vector<std::string> file_paths_{};  // List of database files.
   std::vector<std::string>::iterator current_file_iterator_{};  // Index of file to read from
+  rosbag2_cpp::CompressionMode compression_mode_;
```
Comment (zmichaels11) [inline rosbag2_cpp/include/rosbag2_cpp/readers/sequential_reader.hpp]:
```c++
rosbag2_cpp::CompressionMode compression_mode_{rosbag2_cpp::CompressionMode::NONE};
```

File: rosbag2_cpp/src/rosbag2_cpp/readers/sequential_reader.cpp [hunk 3]
```diff
@@ -18,11 +18,12 @@
 #include <utility>
 #include <vector>
 
-#include "rcpputils/filesystem_helper.hpp"
-
 #include "rosbag2_cpp/logging.hpp"
 #include "rosbag2_cpp/readers/sequential_reader.hpp"
 
+#include "rcpputils/filesystem_helper.hpp"
+#include "rosbag2_compression/zstd_decompressor.hpp"
```
Comment (zmichaels11) [inline rosbag2_cpp/src/rosbag2_cpp/readers/sequential_reader.cpp]: Alphasort and insert a line break after `rcpputils`

File: rosbag2_cpp/include/rosbag2_cpp/compression_options.hpp [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+enum class CompressionMode
+{
+  NONE,
+  FILE,
+  MESSAGE,
+};
+
+CompressionMode compression_mode_from_string(const std::string & compression_mode);
+
+std::string compression_mode_to_string(CompressionMode compression_mode);
```
Comment (zmichaels11) [inline rosbag2_cpp/include/rosbag2_cpp/compression_options.hpp]: Please add unit tests for these two functions

File: rosbag2_cpp/include/rosbag2_cpp/compression_options.hpp [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+#ifndef ROSBAG2_CPP__COMPRESSION_OPTIONS_HPP_
+#define ROSBAG2_CPP__COMPRESSION_OPTIONS_HPP_
+
+#include <string>
+
+namespace rosbag2_cpp
+{
+
+enum class CompressionMode
```
Comment (dabonnie) [inline rosbag2_cpp/include/rosbag2_cpp/compression_options.hpp]: Some doc in this file would be nice.

**Comment by piraka9011** [comment]
Wondering if anyone had thoughts on making the new utility functions `open_storage()` and `setup_compression()` protected vs. private as well as naming.

File: rosbag2_cpp/src/rosbag2_cpp/readers/sequential_reader.cpp [hunk 6]
```diff
@@ -49,6 +49,30 @@ void SequentialReader::reset()
...
+void SequentialReader::open_storage()
+{
+  storage_ = storage_factory_->open_read_only(
+    *current_file_iterator_, metadata_.storage_identifier);
+  if (!storage_) {
+    throw std::runtime_error("No storage could be initialized. Abort");
+  }
+}
+
+void SequentialReader::setup_compression()
```
Comment (dabonnie) [inline rosbag2_cpp/src/rosbag2_cpp/readers/sequential_reader.cpp]: No unit test for this method?

Comment (piraka9011) [reply rosbag2_cpp/src/rosbag2_cpp/readers/sequential_reader.cpp]: Since this is part of the `open()` method and we already check if the proper compression methods are called `decompress_file/message()` then I don't think this needs a unit test.

Comment (piraka9011) [reply rosbag2_cpp/src/rosbag2_cpp/readers/sequential_reader.cpp]: Per hallway discussion, one thing we could do is pass bad compression options and make sure the `open()` method throws.

File: rosbag2_cpp/test/rosbag2_cpp/test_compression_options.cpp [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gmock/gmock.h>
+
+#include <string>
+
+#include "rosbag2_cpp/compression_options.hpp"
+
```
Comment (zmichaels11) [inline rosbag2_cpp/test/rosbag2_cpp/test_compression_options.cpp]: Consider adding unit tests for strings that are lower case/mixed case since you force upper case before comparison in `compression_mode_from_string`

File: rosbag2_cpp/test/rosbag2_cpp/test_compression_options.cpp [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+TEST(CompressionOptionsFromStringTest, BadInputReturnsNoneMode)
+{
+  std::string compression_mode_string{"bad_mode"};
+  auto compression_mode = rosbag2_cpp::compression_mode_from_string(compression_mode_string);
+  EXPECT_EQ(compression_mode, rosbag2_cpp::CompressionMode::NONE);
+}
+
```
Comment (zmichaels11) [inline rosbag2_cpp/test/rosbag2_cpp/test_compression_options.cpp]: Add a unit test where the string is empty

File: rosbag2_cpp/test/rosbag2_cpp/test_compression_options.cpp [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+  EXPECT_EQ(compression_mode, rosbag2_cpp::CompressionMode::FILE);
+}
+
+TEST(CompressionOptionsFromStringTest, MessageStringReturnsMessageMode)
+{
+  std::string compression_mode_string{"MESSAGE"};
+  auto compression_mode = rosbag2_cpp::compression_mode_from_string(compression_mode_string);
+  EXPECT_EQ(compression_mode, rosbag2_cpp::CompressionMode::MESSAGE);
+}
+
```
Comment (zmichaels11) [inline rosbag2_cpp/test/rosbag2_cpp/test_compression_options.cpp]: Add a unit test for `compression_mode_to_string` passed an invalid `CompressionMode` 

I think you can do:
```c++
const auto compression_mode = static_cast<CompressionMode>(100);
```

File: rosbag2_cpp/test/rosbag2_cpp/test_compression_options.cpp [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+#include <gmock/gmock.h>
+
+#include <string>
+
+#include "rosbag2_cpp/compression_options.hpp"
+
+TEST(CompressionOptionsFromStringTest, BadInputReturnsNoneMode)
+{
+  std::string compression_mode_string{"bad_mode"};
+  auto compression_mode = rosbag2_cpp::compression_mode_from_string(compression_mode_string);
```
Comment (zmichaels11) [inline rosbag2_cpp/test/rosbag2_cpp/test_compression_options.cpp]:
```suggestion
const std::string compression_mode_string{"bad_mode"};
const auto compression_mode =
```

File: rosbag2_cpp/src/rosbag2_cpp/compression_options.cpp [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+{
+
+  constexpr const char kCompressionModeNoneStr[] = "NONE";
+  constexpr const char kCompressionModeFileStr[] = "FILE";
+  constexpr const char kCompressionModeMessageStr[] = "MESSAGE";
+
+  std::string to_upper(const std::string & text)
+  {
+    std::string uppercase_text = text;
+    transform(uppercase_text.begin(), uppercase_text.end(), uppercase_text.begin(), ::toupper);
```
Comment (zmichaels11) [inline rosbag2_cpp/src/rosbag2_cpp/compression_options.cpp]: Does `transform` exist in the `std` namespace?
If it does, explicitly call from it
```c++
std::transform(...)
```

File: rosbag2_cpp/src/rosbag2_cpp/compression_options.cpp [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+    return uppercase_text;
+  }
+}  // namespace
+
+CompressionMode compression_mode_from_string(const std::string & compression_mode)
+{
+  auto compression_mode_upper = to_upper(compression_mode);
```
Comment (zmichaels11) [inline rosbag2_cpp/src/rosbag2_cpp/compression_options.cpp]:
```suggestion
const auto compression_mode_upper = to_upper(compression_mode);
```

File: rosbag2_cpp/src/rosbag2_cpp/readers/sequential_reader.cpp [hunk 13]
```diff
@@ -48,6 +49,30 @@ void SequentialReader::reset()
   storage_.reset();
 }
 
+void SequentialReader::open_storage()
+{
+  storage_ = storage_factory_->open_read_only(
+    *current_file_iterator_, metadata_.storage_identifier);
+  if (!storage_) {
+    throw std::runtime_error("No storage could be initialized. Abort");
```
Comment (zmichaels11) [inline rosbag2_cpp/src/rosbag2_cpp/readers/sequential_reader.cpp]: Nit:
```suggestion
    throw std::runtime_error{No storage could be initialized. Abort"};
```

File: rosbag2_cpp/src/rosbag2_cpp/readers/sequential_reader.cpp [hunk 14]
```diff
@@ -112,10 +128,19 @@ bool SequentialReader::has_next()
...
+{
+  if (compression_mode_ == CompressionMode::MESSAGE) {
+    decompressor_->decompress_serialized_bag_message(message);
+  }
+}
+
 std::shared_ptr<rosbag2_storage::SerializedBagMessage> SequentialReader::read_next()
 {
   if (storage_) {
     auto message = storage_->read_next();
```
Comment (dabonnie) [inline rosbag2_cpp/src/rosbag2_cpp/readers/sequential_reader.cpp]: Should message be const?

File: rosbag2_cpp/test/rosbag2_cpp/test_compression_options.cpp [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+{
+  const std::string compression_mode_string{"bad_mode"};
+  auto compression_mode = rosbag2_cpp::compression_mode_from_string(compression_mode_string);
+  EXPECT_EQ(compression_mode, rosbag2_cpp::CompressionMode::NONE);
+}
+
+TEST(CompressionOptionsFromStringTest, EmptyInputReturnsNoneMode)
+{
+  const std::string compression_mode_string;
+  auto compression_mode = rosbag2_cpp::compression_mode_from_string(compression_mode_string);
```
Comment (dabonnie) [inline rosbag2_cpp/test/rosbag2_cpp/test_compression_options.cpp]:
```suggestion
  const auto compression_mode = rosbag2_cpp::compression_mode_from_string(compression_mode_string);
```

File: rosbag2_cpp/test/rosbag2_cpp/test_compression_options.cpp [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+  EXPECT_EQ(compression_mode, rosbag2_cpp::CompressionMode::NONE);
+}
+
+TEST(CompressionOptionsFromStringTest, FileStringReturnsFileMode)
+{
+  std::string compression_mode_string{"file"};
+  auto compression_mode = rosbag2_cpp::compression_mode_from_string(compression_mode_string);
```
Comment (dabonnie) [inline rosbag2_cpp/test/rosbag2_cpp/test_compression_options.cpp]:
```suggestion
  const auto compression_mode = rosbag2_cpp::compression_mode_from_string(compression_mode_string);
```

File: rosbag2_cpp/test/rosbag2_cpp/test_compression_options.cpp [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+  EXPECT_EQ(compression_mode, rosbag2_cpp::CompressionMode::FILE);
+}
+
+TEST(CompressionOptionsFromStringTest, MixedCaseMessageStringReturnsMessageMode)
+{
+  std::string compression_mode_string{"MeSsAgE"};
+  auto compression_mode = rosbag2_cpp::compression_mode_from_string(compression_mode_string);
```
Comment (dabonnie) [inline rosbag2_cpp/test/rosbag2_cpp/test_compression_options.cpp]:
```suggestion
  const auto compression_mode = rosbag2_cpp::compression_mode_from_string(compression_mode_string);
```

File: rosbag2_cpp/test/rosbag2_cpp/test_compression_options.cpp [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+  EXPECT_EQ(compression_mode, rosbag2_cpp::CompressionMode::MESSAGE);
+}
+
+TEST(CompressionOptionsFromStringTest, MessageStringReturnsMessageMode)
+{
+  std::string compression_mode_string{"MESSAGE"};
+  auto compression_mode = rosbag2_cpp::compression_mode_from_string(compression_mode_string);
```
Comment (dabonnie) [inline rosbag2_cpp/test/rosbag2_cpp/test_compression_options.cpp]:
```suggestion
  const auto compression_mode = rosbag2_cpp::compression_mode_from_string(compression_mode_string);
```

File: rosbag2_cpp/test/rosbag2_cpp/test_compression_options.cpp [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+  EXPECT_EQ(compression_mode, rosbag2_cpp::CompressionMode::MESSAGE);
+}
+
+TEST(CompressionOptionsToStringTest, BadModeReturnsNoneString)
+{
+  const auto compression_mode = static_cast<rosbag2_cpp::CompressionMode>(100);
+  auto compression_mode_string = rosbag2_cpp::compression_mode_to_string(compression_mode);
```
Comment (dabonnie) [inline rosbag2_cpp/test/rosbag2_cpp/test_compression_options.cpp]:
```suggestion
  const auto compression_mode_string = rosbag2_cpp::compression_mode_to_string(compression_mode);
```

File: rosbag2_cpp/test/rosbag2_cpp/test_compression_options.cpp [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+  EXPECT_EQ(compression_mode_string, "NONE");
+}
+
+TEST(CompressionOptionsToStringTest, MessageModeReturnsMessageString)
+{
+  auto compression_mode = rosbag2_cpp::CompressionMode::MESSAGE;
```
Comment (dabonnie) [inline rosbag2_cpp/test/rosbag2_cpp/test_compression_options.cpp]:
```suggestion
  const auto compression_mode = rosbag2_cpp::CompressionMode::MESSAGE;
```

File: rosbag2_cpp/test/rosbag2_cpp/test_compression_options.cpp [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+  auto compression_mode_string = rosbag2_cpp::compression_mode_to_string(compression_mode);
```
Comment (dabonnie) [inline rosbag2_cpp/test/rosbag2_cpp/test_compression_options.cpp]:
```suggestion
  const auto compression_mode_string = rosbag2_cpp::compression_mode_to_string(compression_mode);
```

File: rosbag2_cpp/test/rosbag2_cpp/test_compression_options.cpp [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+  EXPECT_EQ(compression_mode_string, "MESSAGE");
+}
+
+TEST(CompressionOptionsToStringTest, FileModeReturnsFileString)
+{
+  auto compression_mode = rosbag2_cpp::CompressionMode::FILE;
```
Comment (dabonnie) [inline rosbag2_cpp/test/rosbag2_cpp/test_compression_options.cpp]:
```suggestion
  const auto compression_mode = rosbag2_cpp::CompressionMode::FILE;
```

File: rosbag2_cpp/test/rosbag2_cpp/test_compression_options.cpp [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+  auto compression_mode_string = rosbag2_cpp::compression_mode_to_string(compression_mode);
```
Comment (dabonnie) [inline rosbag2_cpp/test/rosbag2_cpp/test_compression_options.cpp]:
```suggestion
  const auto compression_mode_string = rosbag2_cpp::compression_mode_to_string(compression_mode);
```

File: rosbag2_cpp/test/rosbag2_cpp/test_compression_options.cpp [hunk 24]
```diff
@@ -0,0 +1,82 @@
...
+  EXPECT_EQ(compression_mode_string, "FILE");
+}
+
+TEST(CompressionOptionsToStringTest, NoneModeReturnsNoneString)
+{
+  auto compression_mode = rosbag2_cpp::CompressionMode::NONE;
```
Comment (dabonnie) [inline rosbag2_cpp/test/rosbag2_cpp/test_compression_options.cpp]:
```suggestion
  const auto compression_mode = rosbag2_cpp::CompressionMode::NONE;
```

File: rosbag2_cpp/test/rosbag2_cpp/test_compression_options.cpp [hunk 25]
```diff
@@ -0,0 +1,82 @@
...
+  auto compression_mode_string = rosbag2_cpp::compression_mode_to_string(compression_mode);
```
Comment (dabonnie) [inline rosbag2_cpp/test/rosbag2_cpp/test_compression_options.cpp]:
```suggestion
  const auto compression_mode_string = rosbag2_cpp::compression_mode_to_string(compression_mode);
```

File: rosbag2_cpp/test/rosbag2_cpp/test_multifile_reader.cpp [hunk 26]
```diff
@@ -39,19 +54,16 @@ class MultifileReaderTest : public Test
...
     storage_serialization_format_("rmw1_format"),
-    relative_path_1_("some_relativ_path_1"),
+    relative_path_1_("some_relative_path_1"),
     relative_path_2_("some_relative_path_2")
   {
-    rosbag2_storage::TopicMetadata topic_with_type;
-    topic_with_type.name = "topic";
-    topic_with_type.type = "test_msgs/BasicTypes";
-    topic_with_type.serialization_format = storage_serialization_format_;
+    auto topic_with_type =
```
Comment (dabonnie) [inline rosbag2_cpp/test/rosbag2_cpp/test_multifile_reader.cpp]: const?

File: rosbag2_cpp/test/rosbag2_cpp/test_multifile_reader.cpp [hunk 27]
```diff
@@ -114,3 +127,95 @@ TEST_F(MultifileReaderTest, get_all_topics_and_types_throws_if_no_storage)
...
+  {
+    decompress_message_called = true;
+  }
+
+  void decompress_file(const std::string &) override
+  {
+    decompress_file_called = true;
+  }
+
+  bool decompress_message_called = false;
```
Comment (dabonnie) [inline rosbag2_cpp/test/rosbag2_cpp/test_multifile_reader.cpp]: Consider using counters instead of flags.

File: rosbag2_cpp/test/rosbag2_cpp/test_multifile_reader.cpp [hunk 28]
```diff
@@ -114,3 +127,95 @@ TEST_F(MultifileReaderTest, get_all_topics_and_types_throws_if_no_storage)
...
+    ON_CALL(*storage_, read_next()).WillByDefault(Return(message));
+  }
+
+  std::shared_ptr<NiceMock<MockStorage>> storage_;
+  std::shared_ptr<NiceMock<MockConverterFactory>> converter_factory_;
+  rosbag2_storage::TopicMetadata topic_metadata_;
+};
+
+TEST_F(ReaderCompressionTest, open_throws_on_bad_compression_format) {
+  auto metadata_io = std::make_unique<NiceMock<MockMetadataIo>>();
```
Comment (dabonnie) [inline rosbag2_cpp/test/rosbag2_cpp/test_multifile_reader.cpp]:
```suggestion
  const auto metadata_io = std::make_unique<NiceMock<MockMetadataIo>>();
```

File: rosbag2_cpp/test/rosbag2_cpp/test_multifile_reader.cpp [hunk 29]
```diff
@@ -114,3 +127,95 @@ TEST_F(MultifileReaderTest, get_all_topics_and_types_throws_if_no_storage)
...
+  void decompress_file(const std::string &) override
+  {
+    decompress_file_called = true;
+  }
+
+  bool decompress_message_called = false;
+  bool decompress_file_called = false;
+};
+
+class ReaderCompressionTest : public Test
```
Comment (dabonnie) [inline rosbag2_cpp/test/rosbag2_cpp/test_multifile_reader.cpp]: There's a lot of local variables that could be const in this class and the following tests.

File: rosbag2_cpp/test/rosbag2_cpp/test_multifile_reader.cpp [hunk 30]
```diff
@@ -114,3 +127,95 @@ TEST_F(MultifileReaderTest, get_all_topics_and_types_throws_if_no_storage)
...
+}
+
+TEST_F(ReaderCompressionTest, decompress_if_metadata_has_file_compression) {
+  auto metadata_io = std::make_unique<NiceMock<MockMetadataIo>>();
+  rosbag2_storage::BagMetadata metadata;
+  metadata.relative_file_paths = {"some_relative_path_1", "some_relative_path_2"};
+  metadata.topics_with_message_count.push_back({{topic_metadata_}, 1});
+  metadata.compression_mode =
+    rosbag2_cpp::compression_mode_to_string(rosbag2_cpp::CompressionMode::FILE);
+  metadata.compression_format = "zstd";
```
Comment (dabonnie) [inline rosbag2_cpp/test/rosbag2_cpp/test_multifile_reader.cpp]: There's no defined zstd string to use?

File: rosbag2_cpp/test/rosbag2_cpp/test_multifile_reader.cpp (hunk 26 above)
Comment (piraka9011) [reply rosbag2_cpp/test/rosbag2_cpp/test_multifile_reader.cpp]: We may want to modify it in a test.

File: rosbag2_cpp/src/rosbag2_cpp/readers/sequential_reader.cpp (hunk 14 above)
Comment (piraka9011) [reply rosbag2_cpp/src/rosbag2_cpp/readers/sequential_reader.cpp]: No, we decompress it in place. It shouldn't be const

File: rosbag2_cpp/test/rosbag2_cpp/test_multifile_reader.cpp (hunk 30 above)
Comment (piraka9011) [reply rosbag2_cpp/test/rosbag2_cpp/test_multifile_reader.cpp]: Nope. There should probably be a definition in the include of the Zstd compressor implementation.

File: rosbag2_cpp/include/rosbag2_cpp/compression_options.hpp [hunk 31]
```diff
@@ -0,0 +1,51 @@
...
+/**
+ * Modes are used to specify whether to compress by individual serialized bag messages or by file.
+ * rosbag2_cpp defaults to NONE.
+ */
+enum class CompressionMode
+{
+  NONE,
+  FILE,
+  MESSAGE,
+};
```
Comment (zmichaels11) [inline rosbag2_cpp/include/rosbag2_cpp/compression_options.hpp]: It might be better to do:
```c++
enum class CompressionMode : uint32_t
{
  NONE = 0,
  FILE,
  MESSAGE,
  LAST_MODE = MESSAGE
};
```
since one of the unit tests is doing a `static_cast<CompressionMode>` on an integer-type.
You can then make the test more robust by doing: 
```c++
auto bad_mode = static_cast<CompressionMode>(static_cast<uint32_t>(CompressionMode::LAST_MODE) + 1);
```

File: rosbag2_cpp/include/rosbag2_cpp/readers/sequential_reader.hpp [hunk 32]
```diff
@@ -121,14 +123,45 @@ class ROSBAG2_CPP_PUBLIC SequentialReader
     const std::string & converter_serialization_format,
     const std::string & storage_serialization_format);
 
+  /**
+   * Opens a storage plugin for read only.
+   *
+   * \throws runtime_error If no storage could be initialized.
```
Comment (zmichaels11) [inline rosbag2_cpp/include/rosbag2_cpp/readers/sequential_reader.hpp]:
```suggestion
   * \throws std::runtime_error If no storage could be initialized.
```

**Comment by piraka9011** [comment]
@ros2/aws-oncall - please run this CI job
Gist: https://gist.githubusercontent.com/piraka9011/8028a6f7bc082936f2057ff5d3fba5c2/raw/fc3f28e4c525ba4f071965076204beaf54298942/ros2.repos
BUILD args: --packages-up-to rosbag2_cpp rosbag2_compression
TEST args: --packages-select rosbag2_cpp rosbag2_compression
Job: ci_launcher

**Comment by prajakta-gokhale** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/8982/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4750/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/7354/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8915/)

**Comment by piraka9011** [comment]
Linux is unstable due to an `rcl` issue.
Windows is complaining that it can't find `rosbag2_cpp::compression_mode_to_string` in `test_multifile_reader`

**Comment by piraka9011** [comment]
@Karsten1987 could you PTAL?

File: rosbag2_cpp/test/rosbag2_cpp/test_multifile_reader.cpp [hunk 33]
```diff
@@ -114,3 +113,95 @@ TEST_F(MultifileReaderTest, get_all_topics_and_types_throws_if_no_storage)
 {
   EXPECT_ANY_THROW(reader_->get_all_topics_and_types());
 }
+
+class FakeSequentialReader : public rosbag2_cpp::readers::SequentialReader
```
Comment (prajakta-gokhale) [inline rosbag2_cpp/test/rosbag2_cpp/test_multifile_reader.cpp]: Maybe use `MockSequentialReader` instead of `FakeSequentialReader` to keep it consistent with rest of the tests..

Comment (piraka9011) [reply rosbag2_cpp/test/rosbag2_cpp/test_multifile_reader.cpp]: Since this is implementing a reader and not mocking it (i.e. using Gmock) it is technically a fake.
Check [this SO](https://stackoverflow.com/questions/346372/whats-the-difference-between-faking-mocking-and-stubbing) answer and this [article](https://www.martinfowler.com/articles/mocksArentStubs.html)

**Comment by dirk-thomas** [comment]
This PR introduced linter warnings which were even reported in the run CI build (see the latest Windows CI). Please fix this asap / by the end of day.

Also in the future do not merge PRs which clearly show regressions in the ran CI build.

**Comment by piraka9011** [comment]
@dirk-thomas Fix in #256.
For context, the build was unstable due to unrelated tests earlier so the uncrustify warnings were probably missed b/c of that.

**Comment by dirk-thomas** [comment]
> For context, the build was unstable due to unrelated tests earlier so the uncrustify warnings were probably missed b/c of that.

The [Windows build](https://ci.ros2.org/job/ci_windows/8925/testReport/) had exact 4 failing tests - all related to this patch. So no unrelated tests obscuring the result in that case.
